src/protections/JoinWaveShortCircuit.ts

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {Protection} from "./IProtection";
+import {Mjolnir} from "../Mjolnir";
+import {NumberProtectionSetting} from "./ProtectionSettings";
+import {LogLevel} from "matrix-bot-sdk";
+import config from "../config";
+
+export const DEFAULT_MAX_PER_HOUR = 50;
+const ONE_HOUR = 3600_000; // 1h in ms
+
+export class JoinWaveShortCircuit extends Protection {
+    requiredStatePermissions = ["m.room.join_rules"]
+
+    private joinBuckets: {
+        [roomId: string]: {
+            lastBucketStart: Date,
+            numberOfJoins: number,
+        }
+    } = {};
+
+    settings = {
+        maxPerHour: new NumberProtectionSetting(DEFAULT_MAX_PER_HOUR)
+    };
+
+    constructor() {
+        super();
+    }
+
+    public get name(): string {
+        return "JoinWaveShortCircuit";
+    }
+
+    public get description(): string {
+        return "If X amount of users join in Y time, set the room to invite-only."
+    }
+
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any) {
+        if (event['type'] !== 'm.room.member') {
+            // Not a join/leave event.
+            return;
+        }
+
+        if (!(roomId in mjolnir.protectedRooms)) {
+            // Not a room we are watching.
+            return;
+        }
+
+        const userId = event['state_key'];
+        if (!userId) {
+            // Ill-formed event.
+            return;
+        }
+
+        const userState = event['content']['membership'];
+        const prevMembership = event['unsigned']?.['prev_content']?.['membership'] || "leave";
+
+        // We look at the previous membership to filter out profile changes
+        if (userState === 'join' && prevMembership !== "join") {
+            // A new join, fallthrough
+        } else {
+            return;
+        }
+
+        // If either the roomId bucket didn't exist, or the bucket has expired, create a new one
+        if (!this.joinBuckets[roomId] || JoinWaveShortCircuit.hasExpired(this.joinBuckets[roomId].lastBucketStart)) {
+            this.joinBuckets[roomId] = {
+                lastBucketStart: new Date(),
+                numberOfJoins: 0
+            }
+        }
```

Comment:
So the bucket will get reset if it's been over an hour since we received the first event in the current bucket?

Reply:
i feel like a token bucket might be a better approach

Reply:
I'm not sure how to efficiently implement that here.

Reply:
Determined via the mjolnir room that i'll do this later.